app/assets/javascripts/covid19/src/webapp/components/structures-table/useStructuresTable.ts

```diff
@@ -0,0 +1,192 @@
+import React, { ReactNode } from "react";
+import { DataGridProps, GridSortModel } from "@material-ui/data-grid";
+import { StructuresTableProps } from "./StructuresTable";
+import { useAppContext } from "../../contexts/app-context";
+import { Covid19Filter, Covid19Info } from "../../../domain/entities/Covid19Info";
+import i18n from "../../../utils/i18n";
+import { isElementOfUnion } from "../../../data/utils/ts-utils";
+
+type Props = StructuresTableProps & {
+    notice: NoticeState;
+    showLoading: () => void;
+    stopLoading: () => void;
+    enableCancellable: () => void;
+};
+
+export function useStructuresTable(props: Props) {
+    const {
+        search,
+        setSearch: setSearch0,
+        notice,
+        showLoading,
+        stopLoading,
+        enableCancellable,
+    } = props;
+
+    const { compositionRoot } = useAppContext();
+
+    //data
+    const [data, setData] = React.useState<Covid19Info>({
+        count: 0,
+        structures: [],
+        validationSources: [],
+    });
+
+    //paging
+    const [page, setPage] = React.useState(0);
+    const pageSize = React.useRef(pageSizes[0] ?? 10);
+
+    //sort filter
+    const [sortModel, setSortModel] = React.useState<GridSortModel>(noSort);
+    const [filterState, setFilterState] = React.useState(initialFilterState);
+
+    //cancel get request
+    const cancelLoadDataRef: React.MutableRefObject<SelfCancellable> = React.useRef(() => {});
+
+    const setSearch = React.useCallback(
+        (value: string) => {
+            // ANALYTICS TO BE REBUILD
+            // sendAnalytics("search", {
+            //     on: "covid_table",
+            //     query: value,
+            // });
+            setSearch0(value.toLowerCase());
+            setSortModel(sort => (value ? noSort : sort));
+        },
+        [setSearch0]
+    );
+
+    const resetPageAndSorting = React.useCallback<GridProp<"onSortModelChange">>(modelParams => {
+        setSortModel(modelParams.sortModel);
+    }, []);
+
+    const getData = React.useCallback(
+        (page: number, pageSize: number, onSuccess?: () => void) => {
+            showLoading();
+            if (pageSize > 25) enableCancellable();
+            const sortOption = sortModel[0];
+
+            const sort =
+                sortOption && isElementOfUnion(sortOption.field, sortingFields) && sortOption.sort
+                    ? {
+                          field: sortOption.field,
+                          order: sortOption.sort,
+                      }
+                    : ({ field: "releaseDate", order: "desc" } as const);
+
+            const cancelGetData = compositionRoot.getCovid19Info
+                .execute({ page, pageSize, filter: filterState, sort, query: search })
+                .bitap(stopLoading, stopLoading)
```

Comment:
Move stopLoading to both branches of run. The idea is not to any side-effect until run.